src/qibolab/compilers/compiler.py

```diff
@@ -0,0 +1,171 @@
+from collections import defaultdict
+from dataclasses import dataclass, field
+
+import numpy as np
+from qibo import gates
+from qibo.config import raise_error
+from qibo.states import CircuitResult
+
+from qibolab.compilers.default import (
+    cz_rule,
+    identity_rule,
+    measurement_rule,
+    rz_rule,
+    u3_rule,
+    z_rule,
+)
+from qibolab.pulses import PulseSequence, ReadoutPulse
+
+
+class MeasurementMap(dict):
+    """Map from each measurement gate to the sequence of readout pulses implementing it."""
+
+    def __call__(self, backend, circuit, readout, nshots):
+        """Assign measurement outcomes to :class:`qibo.states.MeasurementResult` for each gate.
+
+        Args:
+            backend (:class:`qibo.backends.abstract.AbstractBackend`): Backend object to
+                be assigned in the result object.
+            circuit (:class:`qibo.models.Circuit`): Circuit object that the measurement map
+                was produced for. Needed
+            readout (dict): Dictionary containing acquisition results (:class:`qibolab.results.ExecutionResults`)
+                and shot values for the measurements performed on hardware.
+            nshots (int): Number of shots performed during the circuit execution.
+
+        Returns:
+            :class:`qibo.states.CircuitResult` object containing the results acquired from the circuit execution.
+        """
+        result = CircuitResult(backend, circuit, readout, nshots)
+        for gate, sequence in self.items():
+            samples = []
+            for pulse in sequence.pulses:
+                shots = readout[pulse.serial].shots
+                if shots is not None:
+                    samples.append(shots)
```

Comment:
Just an alternative, you might not like:
```suggestion
            _samples = (readout[pulse.serial].shots for pulse in sequence.pulses)
            samples = list(filter(lambda x: x is not None, _samples))
```

I like because it sounds more like I would pronounce it (personal opinion).

(It would have been even nicer if the filtering were on `pulse`, rather than on `shots`, in that case you would have only needed a single comprehension)

Reply:
I believe we can also replace the `for` loop with a `map` in `_samples` (implemented).